synapseclient/api/file_services.py

```diff
@@ -0,0 +1,335 @@
+"""This module is responsible for exposing the services defined at:
+<https://rest-docs.synapse.org/rest/#org.sagebionetworks.repo.web.controller.EntityController>
+"""
+
+import json
+import mimetypes
+import os
+from typing import Any, Dict, Optional, List, Union, TYPE_CHECKING
+
+
+from synapseclient.core import utils
+from synapseclient.core.constants import concrete_types
+from synapseclient.api.entity_services import get_upload_destination
+
+if TYPE_CHECKING:
+    from synapseclient import Synapse
+
+
+async def post_file_multipart(
+    upload_request_payload: Dict[str, Any],
+    force_restart: bool,
+    endpoint: str,
+    synapse_client: Optional["Synapse"] = None,
+) -> Dict[str, str]:
+    """
+    <https://rest-docs.synapse.org/rest/POST/file/multipart.html>
+
+    Arguments:
+        upload_request_payload: The request matching
+            <https://rest-docs.synapse.org/rest/org/sagebionetworks/repo/model/file/MultipartRequest.html>
+        force_restart: Optional parameter. When True, any upload state for the given
+            file will be cleared and a new upload will be started.
+        endpoint: Server endpoint to call to.
+        synapse_client: If not passed in or None this will use the last client from
+            the `.login()` method.
+
+    Returns:
+        The requested multipart upload status matching
+            <https://rest-docs.synapse.org/rest/org/sagebionetworks/repo/model/file/MultipartUploadStatus.html>
+    """
+    from synapseclient import Synapse
+
+    client = Synapse.get_client(synapse_client=synapse_client)
+    return await client.rest_post_async(
+        f"/file/multipart?forceRestart={str(force_restart).lower()}",
+        json.dumps(upload_request_payload),
+        endpoint=endpoint,
+    )
+
+
+async def put_file_multipart_add(
+    upload_id: str,
+    part_number: int,
+    md5_hex: str,
+    synapse_client: Optional["Synapse"] = None,
+) -> Dict[str, Any]:
+    """
+    <https://rest-docs.synapse.org/rest/PUT/file/multipart/uploadId/add/partNumber.html>
+
+    Arguments:
+        upload_id: The unique identifier of the file upload.
+        part_number: The part number to add. Must be a number between 1 and 10,000.
```

Comment:
Do you think we should enforce this limit (1-10,000) in the code itself?

Reply:
At this point I do not think it's needed - But I do think a larger-scale input validation is going to be important in the Open API generated client.